packages/cardano-services-client/test/HttpProvider.test.ts

```diff
@@ -0,0 +1,175 @@
+import express, { RequestHandler } from 'express';
+
+type ComplexArg2 = { map: Map<string, Buffer> };
+type ComplexResponse = Map<bigint, Buffer>[];
+interface TestProvider {
+  noArgsEmptyReturn(): Promise<void>;
+  complexArgsAndReturn(arg1: bigint, arg2: ComplexArg2): Promise<ComplexResponse>;
+}
+
+const createStubHttpProviderServer = async (port: number, path: string, handler: RequestHandler) => {
+  const app = express();
+  app.use(express.json());
+  app.post(path, handler);
+  const server = await new Promise<Server>((resolve) => {
+    const result = app.listen(port, () => resolve(result));
+  });
+  return () => new Promise((resolve) => server.close(resolve));
+};
+
+const stubProviderPaths = {
+  complexArgsAndReturn: '/complex',
+  noArgsEmptyReturn: '/simple'
+};
+
+describe('createHttpServer', () => {
+  let port: number;
+  let baseUrl: string;
+
+  const createTxSubmitProviderClient = (
+    config: Pick<HttpProviderConfig<TestProvider>, 'gotOptions' | 'mapError'> = {}
+  ) =>
+    createHttpProvider<TestProvider>({
+      baseUrl,
+      paths: stubProviderPaths,
+      ...config
+    });
+
+  beforeAll(async () => {
+    port = await getPort();
+    baseUrl = `http://localhost:${port}`;
+  });
+
+  it('attempting to access unimplemented method throws ProviderError', async () => {
+    const provider = createTxSubmitProviderClient();
+    expect(() => (provider as any).doesntExist).toThrowError(ProviderError);
+  });
+
+  describe('method with no args and void return', () => {
+    it('calls http server with empty args in req body and parses the response', async () => {
+      const provider = createTxSubmitProviderClient();
+      const closeServer = await createStubHttpProviderServer(port, stubProviderPaths.noArgsEmptyReturn, (req, res) => {
+        expect(req.body.args).toEqual([]);
+        res.send();
+      });
+      const response = await provider.noArgsEmptyReturn();
+      expect(response).toBe(undefined);
+      await closeServer();
+    });
+  });
+
+  describe('method with complex args and return', () => {
+    it('serializes args and deserializes response using core serializableObject', async () => {
+      const arg1 = 123n;
+      const arg2: ComplexArg2 = { map: new Map([['key', Buffer.from('abc')]]) };
+      const expectedResponse: ComplexResponse = [new Map([[1234n, Buffer.from('response data')]])];
+      const provider = createTxSubmitProviderClient();
+      const closeServer = await createStubHttpProviderServer(
+        port,
+        stubProviderPaths.complexArgsAndReturn,
+        (req, res) => {
+          expect(util.fromSerializableObject(req.body.args)).toEqual([arg1, arg2]);
+          res.send(util.toSerializableObject(expectedResponse));
+        }
+      );
+      const response = await provider.complexArgsAndReturn(arg1, arg2);
+      expect(response).toEqual(expectedResponse);
+      await closeServer();
+    });
+  });
+
+  it('passes through got options', async () => {
```

Comment:
Not strictly necessary, but this and a few more tests in this file could use `expect.assertions(n)`. This came from my commits